// @sap/cds/lib/srv/protocols/http.js
const cds = require('../../index'), { decodeURI } = cds.utils
const express = require('express')
const production = process.env.NODE_ENV === 'production'
const restrict_all = cds.env.requires.auth?.restrict_all_services !== false
const restricted_by_default = production && restrict_all ? ['authenticated-user'] : false


class HttpAdapter {

  /** Constructs and returns a new express.Router */
  constructor (srv,o={}) {
    this.kind = o.kind || this.constructor.name.replace(/Adapter$/,'').toLowerCase()
    this.service = srv
    this.options = o
    return this.router //> constructed by getter
  }

  /** The actual Router factory. Subclasses override this to add specific handlers. */
  get router() {
    let router = super.router = (new express.Router) .use (this.http_log.bind(this))
    let assert_roles = this.requires_check()
    if (assert_roles) router.use (assert_roles)
    return router
  }

  /** Handler to log all incoming requests */
  http_log (r,_,next) {
    this.logger = cds.log(this.kind)
    this.log(r)
    next()
  }

  /** Subclasses may override this method to log incoming requests. */
  log (req, LOG = this.logger) { LOG._info && LOG.info (
    req.method,
    decodeURI (req.baseUrl + req.path),
    Object.keys (req.query).length ? { ...req.query } : ''
  )}

  /** Returns a handler to check required roles, or null if no check required. */
  requires_check() {
    const d = this.service.definition
    const roles = d['@requires'] || d['@restrict']?.map(r => r.to).flat().filter(r => r)
    const required = !roles?.length ? restricted_by_default : Array.isArray(roles) ? roles : [roles]
    if (required) return function requires_check (req, res, next) {
      const user = cds.context.user
      if (required.some(role => user.has(role))) return next()
      else if (user._is_anonymous) return next(401) // request login
      else throw Object.assign(new Error, { code: 403, reason: `User '${user.id}' is lacking required roles: [${required}]`, user, required })
    }
  }
}


module.exports = HttpAdapter
